File: ios/Khrysalis/src/net/ConnectedWebSocket.actual.swift
import RxSwift
import RxRelay
import Starscream

//--- ConnectedWebSocket.{
public final class ConnectedWebSocket: WebSocketDelegate, Disposable {
    
    
    //--- ConnectedWebSocket.ownConnection
    public var ownConnection = PublishSubject<ConnectedWebSocket>.create()
    var underlyingSocket: WebSocket?
    var url: String
    //--- ConnectedWebSocket.read
    private let _read: PublishSubject<WebSocketFrame> = PublishSubject.create()
    public var read: Observable<WebSocketFrame> { return _read }
    
    //--- ConnectedWebSocket.Primary Constructor
    init(url: String) {
        self.url = url
    }
    
    
    //--- ConnectedWebSocket.onComplete()
    public func onComplete() -> Void {
        underlyingSocket?.disconnect(closeCode: 1000)
    }
    
    //--- ConnectedWebSocket.onNext(WebSocketFrame)
    public func onNext(_ t: WebSocketFrame) -> Void {
        if let text = t.text {
            underlyingSocket?.write(string: text, completion: nil)
        }
        if let binary = t.binary {
            underlyingSocket?.write(data: binary, completion: nil)
        }
    }
    public func onNext(t: WebSocketFrame) -> Void {
        return onNext(t)
    }
    
    //--- ConnectedWebSocket.onError(Throwable)
    public func onError(_ e: Error) -> Void {
        underlyingSocket?.disconnect(closeCode: 1011)
    }
    public func onError(e: Error) -> Void {
        return onError(e)
    }
    
    //--- Receive
    public func didReceive(event: WebSocketEvent, client: WebSocket) {
        switch event {
        case .binary(let data):
            print("Socket to \(url) got binary message of length '\(data.count)'")
            _read.onNext(WebSocketFrame(binary: data))
            break
        case .text(let string):
            print("Socket to \(url) got message '\(string)'")
            _read.onNext(WebSocketFrame(text: string))
            break
        case .connected(let headers):
            print("Socket to \(url) opened successfully.")
            ownConnection.onNext(self)
            break
        case .disconnected(let reason, let code):
            print("Socket to \(url) disconnecting with code \(code). Reason: \(reason)")
            ownConnection.onCompleted()
            _read.onComplete()
            break
        case .error(let error):
            print("Socket to \(url) failed with error \(error)")
            ownConnection.onError(error ?? Exception())
            read.onError(error ?? Exception())
            break
        case .cancelled:
            print("Socket to \(url) cancelled")
            ownConnection.onError(Exception("Socket connection cancelled."))
            _read.onComplete()
            break
        default:
            break
        }
    }
    
    public func dispose() {
        print("Socket to \(url) was disposed, closing with OK code.")
        underlyingSocket?.disconnect(closeCode: 1000)
        ownConnection.onCompleted()
        _read.onComplete()
    }
    
    //--- ConnectedWebSocket.}
}

extension WebSocketFrame: CustomStringConvertible {
    public var description: String {
        return text ?? "<binary>"
    }
}

File: android/src/main/java/com/lightningkite/khrysalis/net/ConnectedWebSocket.actual.kt
package com.lightningkite.khrysalis.net

import android.util.Log
import com.lightningkite.khrysalis.PlatformSpecific
import com.lightningkite.khrysalis.post
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.subjects.PublishSubject
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString

class ConnectedWebSocket(val url: String) : WebSocketListener(),
    Observer<WebSocketFrame> {
    internal var underlyingSocket: WebSocket? = null
    private val _read =
        PublishSubject.create<WebSocketFrame>()
    val ownConnection =
        PublishSubject.create<ConnectedWebSocket>()
    val read: Observable<WebSocketFrame> = _read
    var justStarted = false

    @PlatformSpecific
    override fun onOpen(webSocket: WebSocket, response: Response) {
        justStarted = true
        post {
            println("Socket to $url opened successfully.")
            ownConnection.onNext(this)
            post {
                justStarted = false
            }
        }
    }

    @PlatformSpecific
    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
        post {
            try {
                println("Socket to $url failed with $t.")
                ownConnection.onError(t)
                _read.onError(t)
            } catch (e: Exception) {
                Log.e("ConnectedWebSocket", "Failed to deliver error")
                e.printStackTrace()
            }
        }
    }

    @PlatformSpecific
    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
        post {
            println("Socket to $url closing.")
            ownConnection.onComplete()
            _read.onComplete()
        }
    }

    @PlatformSpecific
    override fun onMessage(webSocket: WebSocket, text: String) {
        //TODO: You can measure the jank of an Android application by the number of nested posts.  There's TWO here.  EWW.
        post {
            if(justStarted){
                post {
                    println("Socket to $url got message '$text'.")
                    _read.onNext(WebSocketFrame(text = text))
                }
            } else {
                println("Socket to $url got message '$text'.")
                _read.onNext(WebSocketFrame(text = text))
            }
        }
    }

    @PlatformSpecific
    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
        post {
            println("Socket to $url got binary message of length ${bytes.size()}.")
            _read.onNext(WebSocketFrame(binary = bytes.toByteArray()))
        }
    }

    @PlatformSpecific
    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
        println("Socket to $url closed.")
    }

    override fun onComplete() {
        underlyingSocket?.close(1000, null)
    }

    @PlatformSpecific
    override fun onSubscribe(d: Disposable) {
    }

    override fun onNext(t: WebSocketFrame) {
        t.text?.let {
            underlyingSocket?.send(it)
        }
        t.binary?.let { binary ->
            underlyingSocket?.send(ByteString.of(binary, 0, binary.size))
        }
    }

    override fun onError(e: Throwable) {
        underlyingSocket?.close(1011, e.message)
    }
}